### PR TITLE
Optionally deny floating point operations

### LINKED ISSUE #49: Add float ban as optional validation check

### PR DESCRIPTION
**Pull request by Vurich**
Closes #49

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -431,6 +430,55 @@ impl Module {
 		})
 	}
 
+	/// Create `Module` from `parity_wasm::elements::Module`.
```
Comment (pepyakin) [inline src/lib.rs]: These seems to be not relevant. Let's update this to actually reflect the method's intention

File: src/validation/mod.rs [hunk 2]
```diff
@@ -50,6 +50,97 @@ impl ::std::ops::Deref for ValidatedModule {
...
+			use parity_wasm::elements::Opcode::*;
+
+			macro_rules! match_eq {
+                    ($pattern:pat) => {
+                        |val| if let $pattern = *val { true } else { false }
+                    }
+                }
+
+			const DENIED: &[fn(&Opcode) -> bool] = &[
+				match_eq!(F32Load(_, _)),
```
Comment (pepyakin) [inline src/validation/mod.rs]: nice!

File: src/validation/mod.rs [hunk 3]
```diff
@@ -50,6 +50,97 @@ impl ::std::ops::Deref for ValidatedModule {
...
+				match_eq!(I64TruncSF32),
+				match_eq!(I64TruncUF32),
+				match_eq!(I64TruncSF64),
+				match_eq!(I64TruncUF64),
+				match_eq!(I32ReinterpretF32),
+				match_eq!(I64ReinterpretF64),
+			];
+
+			if DENIED.iter().any(|is_denied| is_denied(op)) {
+				return Err(Error(format!("Unimplemented opcode: {:?}", op)));
```
Comment (pepyakin) [inline src/validation/mod.rs]: How about returning something like more specific, like "usage of a forbidden opcode"?

**Comment by pepyakin** [comment]
the thing that worries me is floats are still possible to use, e.g.:

```WebAssembly
(module
   (func (export "derp") (param f32) (result f32)
     get_local 0     
   )
)
```

I didn't found a case where it would be a problem so far, at least in our specific use cases (i.e kovan-wasm). @NikVolf what do you think ?

File: src/lib.rs (hunk 1 above)
Comment (Vurich) [reply src/lib.rs]: Ah crap, I copied it from another method

**Comment by Vurich** [comment]
@pepyakin Isn't that denied by denying `F32Load`?

**Comment by NikVolf** [comment]
Floats are already disabled by gas metering instrumentation in kovan-wasm, so feel free to postulate your own requirements for wasmi @pepyakin

**Comment by pepyakin** [comment]
No, that code will be allowed (will pass the `deny_floating_point` check) since it doesn't contain `f32.load` (or any other fp specific) operations: instructions such as `get_local`/`get_global` etc are generic and their type depends on the location they refer to.

In my understanding, without any floating point operations there is nothing the user can do with these values, only juggling with them with `get_local`/`get_global`/`call`/etc, this should be ok but i'm not 100% sure.

> Floats are already disabled by gas metering instrumentation in kovan-wasm

Ah, indeed, this isn't currently a problem in kovan-wasm and I don't yet see any further uses of float ban. Do we actually need one in `wasmi`?

**Comment by Vurich** [comment]
Isn't the reason that we want to deny floating point to avoid differences in rounding? I don't think it's a problem to allow `get_{global,local}`. I guess we can allow `{f32,f64}.load` too

**Comment by sunfishcode** [comment]
WebAssembly scalar floating point has completely deterministic rounding. The parts that are nondeterministic are the bits of NaN values returned from arithmetic instructions, and support for subnormal values in proposed SIMD arithmetic instructions.

**Comment by pepyakin** [comment]
Hm, I had an impression that we want to deny floating point ops to avoid differences in bit patterns of `NaN` and I wasn't aware of anything regarding to rounding. The [spec states](https://webassembly.github.io/spec/core/exec/numerics.html#rounding) that the rounding is fully defined operation (am I correct?)

And operations with NaN values aren't, their payload might be different on the input values and the exact operation. If you store this value with `fN.store` (or somehow observe their bitpattern, i guess  `copysign`?) you can have non deterministic result thus consensus problems.

**Pull request merged**